dbos-rules.ts

```diff
@@ -0,0 +1,435 @@
+      if (ignoreAwaitsForCallsWithAContextParam && validTypeExistsInFunctionCallParams(functionCall, validSet)) {
+       return;
+      }
+
+    return "awaitingOnNotAllowedType";
+    }
+  }
+}
+
+////////// This is the main function that recurs on the `ts-morph` AST
+
+function evaluateFunctionForDeterminism(fn: FunctionOrMethod) {
+  const body = fn.getBody();
+
+  if (body === undefined) {
+    throw new Error("When would a function not have a body?");
+  }
+
+  const stack: Set<string>[] = [new Set()];
+  const getCurrentFrame = () => stack[stack.length - 1];
+  const pushFrame = () => stack.push(new Set());
+  const popFrame = () => stack.pop();
+  const isLocal = (name: string) => stack.some((frame) => frame.has(name));
+
+  const detCheckers: DetChecker[] = [mutatesGlobalVariable, callsBannedFunction, awaitsOnNotAllowedType];
+
+  function checkNodeForGlobalVarUsage(node: Node) {
+    const locals = getCurrentFrame();
+
+    if (Node.isClassDeclaration(node)) {
+      evaluateClassForDeterminism(node);
+      return;
+    }
+    else if (Node.isFunctionDeclaration(node)) { // || Node.isArrowFunction(node)) {
+      /* Not checking if this function should be deterministic
+      strictly, since it might have nondeterministic subfunctions */
+      evaluateFunctionForDeterminism(node);
+      return;
+    }
+    else if (Node.isBlock(node)) {
+      pushFrame();
+      node.forEachChild(checkNodeForGlobalVarUsage);
+      popFrame();
+      return;
+    }
+    // Note: parameters are not considered to be locals here (modifying them is not allowed, currently!)
+    else if (Node.isVariableDeclaration(node)) {
+      locals.add(node.getName());
+    }
+    else if (functionShouldBeDeterministic(fn)) {
+
+      detCheckers.forEach((detChecker) => {
+        const messageKey = detChecker(node, fn, isLocal);
+
+        if (messageKey !== undefined) {
+          const correspondingEslintNode = makeEslintNode!(node);
+          GLOBAL_TOOLS!.eslintContext.report({ node: correspondingEslintNode, messageId: messageKey });
+        }
+      });
+
+      // console.log(`Not accounted for (det function, ${node.getKindName()})... (${node.print()})`);
+    }
+    else {
+      // console.log("Not accounted for (nondet function)...");
+    }
+
+    node.forEachChild(checkNodeForGlobalVarUsage);
+  }
+
+  body.forEachChild(checkNodeForGlobalVarUsage);
+}
+
+////////// This is the entrypoint for running the determinism analysis with `ts-morph`
+
+function analyzeRootNodeForDeterminism(eslintNode: EslintNode, eslintContext: EslintContext) {
+  const parserServices = ESLintUtils.getParserServices(eslintContext);
+
+  GLOBAL_TOOLS = {
+    eslintContext: eslintContext,
+    parserServices: parserServices,
+    typeChecker: parserServices.program.getTypeChecker()
+  };
+
+  const tsMorphNode = makeTsMorphNode(eslintNode);
+
+  try {
+    if (Node.isSourceFile(tsMorphNode)) {
+      tsMorphNode.getFunctions().forEach(evaluateFunctionForDeterminism);
```

Comment:
Since functions outside of classes can't have decorators, do we still need this?

Reply:
Yes - there might be classes inside of the functions, and the `getFunctions` and `getClasses` calls only give you the outermost classes and functions, so you have to recur on the functions and classes inside of all the functions and classes respectively